lumicks/pylake/detail/imaging_mixins.py

```diff
@@ -1,11 +1,112 @@
+            The data type of a single color channel in the resulting image.
+        clip : bool
+            If enabled, the image data will be clipped to fit into the desired `dtype`. This option
+            is disabled by default: an error will be raise if the data does not fit.
+        """
+        # If the the exported tiff should be casted to `dtype`, get the full image stack to later
+        # safely cast it. Otherwise, try to get an iterator, to save memory.
+        frames = self._tiff_frames(iterator=dtype is None)
+        timestamp_ranges = self._tiff_timestamp_ranges()
+
+        # Check length of timestamp ranges, as this is easier than checking the number of frames,
+        # which could be an iterator or an np.ndarray.
+        if len(timestamp_ranges) == 0:
+            raise RuntimeError("Can't export TIFF if there are no images.")
+
+        # Cast frames if a specific dtype is requested.
+        def _cast_image(image, dtype=np.float32, clip=False):
+            # Check if requested dtype can fit image values without an overflow
+            info = np.finfo(dtype) if np.dtype(dtype).kind == "f" else np.iinfo(dtype)
+            if not clip and (np.min(image) < info.min or np.max(image) > info.max):
+                raise RuntimeError(
+                    f"Can't safely export image with `dtype={dtype.__name__}` channels."
+                    f" Switch to a larger `dtype` in order to safely store everything"
+                    f" or pass `force=True` to clip the data."
+                )
+
+            return image.astype(dtype)
+
+        frames = _cast_image(frames, dtype=dtype, clip=clip) if dtype else frames
+
+        def extratags(timestamp_range):
+            """Create the extratags tuple used for the `TiffWriter.write()` method
+
+            Notes
+            -----
+            See TIFF specification for the definition of tags (or fields)
+            """
+            if timestamp_range is None:
+                return None
+            # DateTime, str, len, start:stop
+            datetime = f"{timestamp_range[0]}:{timestamp_range[1]}"
+            datetime = (306, "s", len(datetime), datetime)
+            return (datetime,)
+
+        # Get TIFF image description field
+        description = json.dumps(self._tiff_image_metadata(), indent=4)
+
+        # Get kwargs for the TiffWriter.write() method
+        tiff_write_kwargs = self._tiff_writer_kwargs()
+
+        # Save the tiff file page by page
+        with tifffile.TiffWriter(filename) as tif:
+            for frame, timestamp_range in zip(frames, timestamp_ranges):
+                tif.write(
+                    frame,
+                    contiguous=False,  # write tags on each page
+                    extratags=extratags(timestamp_range),
+                    metadata=None,  # suppress tifffile default ImageDescription tag
+                    description=description,
+                    **tiff_write_kwargs,
+                )
```

Comment:
I wanted to avoid recalling the methods `_tiff_image_metadata()` and `_tiff_writer_kwargs()` for every single frame. But the performance impact is small, so OK -> changed.

Reply:
ah good point, didn't think of that. but if the performance impact is negligible then I guess it's ok